src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs

```diff
@@ -516,6 +524,113 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
         throw new NotImplementedException($"An unsupported authentication method has been configured: {Options.AuthenticationMethod}");
     }
 
+    private bool GetConfigString(string name, [NotNullWhen(true)] out string? value)
+    {
+        if (_configuration?.AdditionalData.TryGetValue(name, out var configValue) ?? false)
+        {
+            if (configValue is string v)
+            {
+                value = v;
+                return true;
+            }
+        }
+        value = null;
+        return false;
+    }
+
+    private bool ConfigFlagEnabled(string name)
+    {
+        if (_configuration?.AdditionalData.TryGetValue(name, out var configValue) ?? false)
+        {
+            if (configValue is bool v)
+            {
+                return v;
+            }
+        }
+        return false;
+    }
+
+    private async Task PushAuthorizationRequest(OpenIdConnectMessage authorizeRequest, AuthenticationProperties properties)
+    {
+        if (GetConfigString("pushed_authorization_request_endpoint", out var parEndpoint))
+        {
+            // Build context and run event
+            var parRequest = new HttpRequestMessage(HttpMethod.Post, parEndpoint)
+            {
+                Content = new FormUrlEncodedContent(authorizeRequest.Parameters),
+            };
+            var context = new PushedAuthorizationContext(Context, Scheme, Options, authorizeRequest, parRequest, properties);
+            await Events.PushAuthorization(context);
```

Comment:
I just noticed that the `authorizeRequest.Parameters` are already serialized into the `FormUrlEncodedContent` by the time this event is called. I know that's probably because you want to allow the event to modify the `HttpRequestMessage`, but I think we should set `HttpRequestMessage.Content` after this event if it's still null.

Reply:
Instead of allowing the event to modify the request message, I'm just passing an `OpenIdConnectMessage` that the handler later pushes. With no message, we don't have to worry about the message content being null.

The context now can affect the flow of control in the handler in 3 ways:
- If it handles authentication with `HandleClientAuthentication`, we skip setting the client secret
- If it decides that we don't need to push with `SkipPush`, we completely skip pushing, and redirect as normal
- If it decides to completely control the call to the par endpoint with `HandlePush`, the event won't make the par request and instead will use the `request_uri ` passed to `HandlePush`
- The event can also tweak parameters before they are pushed, but not call any of the context's methods to get otherwise default behavior.

The downside to this is that it is possible to use the API in an ambiguous way: what should the handler do if the event calls both `HandlePush` and `SkipPush`? Should that be an error? A logged warning? Should the last call take precedence?

Reply:
> The downside to this is that it is possible to use the API in an ambiguous way: what should the handler do if the event calls both `HandlePush` and `SkipPush`? Should that be an error? A logged warning? Should the last call take precedence?

I think throwing an `InvalidOperationException` is a good choice if we detect this. I don't think it should depend on order.